agent/xds/builtinextensions/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,167 @@
+package ratelimit
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	envoy_cluster_v3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	envoy_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoy_listener_v3 "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	envoy_route_v3 "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	envoy_ratelimit "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/local_ratelimit/v3"
+	envoy_http_v3 "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	envoy_type_v3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	envoy_resource_v3 "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"github.com/hashicorp/go-multierror"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"github.com/hashicorp/consul/agent/xds/builtinextensiontemplate"
+	"github.com/hashicorp/consul/agent/xds/xdscommon"
+	"github.com/hashicorp/consul/api"
+)
+
+type ratelimit struct {
+	ProxyType     string
+	Listener      string
+	MaxTokens     int
+	TokensPerFill int
+	FillInterval  int
+}
+
+var _ builtinextensiontemplate.Plugin = (*ratelimit)(nil)
+
+// MakeRatelimitExtension is a builtinextensiontemplate.PluginConstructor for a builtinextensiontemplate.EnvoyExtension.
+func MakeRatelimitExtension(ext xdscommon.ExtensionConfiguration) (builtinextensiontemplate.Plugin, error) {
+	var resultErr error
+	var plugin ratelimit
+
+	if name := ext.EnvoyExtension.Name; name != api.BuiltinRatelimitExtension {
+		return nil, fmt.Errorf("expected extension name 'ratelimit' but got %q", name)
+	}
+
+	if err := mapstructure.Decode(ext.EnvoyExtension.Arguments, &plugin); err != nil {
+		return nil, fmt.Errorf("error decoding extension arguments: %v", err)
+	}
+
+	if plugin.FillInterval == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("fillInterval is required"))
+	}
+
+	if plugin.MaxTokens == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("MaxTokens must be greater than 0"))
+	}
+
+	if plugin.TokensPerFill == 0 {
+		resultErr = multierror.Append(resultErr, fmt.Errorf("TokensPerFill must be greater than 0"))
+	}
```

Comment:
According to envoy's docs `TokensPerFill` defaults to 1 when not specified. Do we want this field explicit in Consul anyways?